### PR TITLE
feat(rome_service): traverse upwards the fs to discover the config file

### PR DESCRIPTION
**Pull request by ematipico**
## Summary

This PR adds a way to discover `rome.json` by traversing the file system upwards. 

## Test Plan

I manually tested the feature using the CLI in our internal repository, and tested manually the VSCode extension.

Below a recording to demonstrate the feature:

https://user-images.githubusercontent.com/602478/220338807-cb241498-807a-4bba-a8d6-2b1644e17d1e.mov

## Documentation

- [x] The PR requires documentation
- [x] I will create a new PR to update the documentation

### CODE REVIEW

**Comment by ematipico** [comment]
CI is failing on ubuntu... it seems the some code hits an infinite loop

**Comment by realtimetodie** [comment]
I like this idea and I think this is useful, nonetheless I would appreciate it if this could become an optional opt-in feature, for example with a flag `--auto-search`. Traversing upwards breaks sandboxing in Bazel. This is a problem. For example, a rome.json configuration file in a parent directory would create false positives.

File: crates/rome_service/src/configuration/mod.rs [hunk 1]
```diff
@@ -129,53 +133,85 @@ pub enum BasePath {
...
+        return match file {
+            Ok(mut file) => {
+                let mut buffer = String::new();
+                file.read_to_string(&mut buffer).map_err(|_| {
+                    WorkspaceError::cant_read_file(format!("{}", configuration_path.display()))
+                })?;
+
+                let deserialized = deserialize_from_json::<Configuration>(&buffer)
+                    .with_file_path(&configuration_path.display().to_string());
+                Ok(Some(deserialized))
```
Comment (realtimetodie) [inline crates/rome_service/src/configuration/mod.rs:189]:
```suggestion
                if from_parent {
                  // Display a message that the rome.json configuration file was found using upwards-traversal (auto-search)
                }
                Ok(Some(deserialized))
```

File: crates/rome_service/src/configuration/mod.rs [hunk 2]
```diff
@@ -129,53 +133,85 @@ pub enum BasePath {
...
-                "Could not read the configuration file from {:?}, reason:\n {}",
-                configuration_path.display(),
-                err
-            );
-            Ok(None)
-        }
+            Err(err) => {
+                if let Some(path) = configuration_path.parent() {
+                    if path.is_dir() {
+                        configuration_path = path.join(config_name);
```
Comment (realtimetodie) [inline crates/rome_service/src/configuration/mod.rs:194]:
```suggestion
                        from_parent = true;
                        configuration_path = path.join(config_name);
```

**Comment by ematipico** [comment]
> I like this idea and I think this is useful, nonetheless I would appreciate it if this could become an optional opt-in feature, for example with a flag `--auto-search`. Traversing upwards breaks sandboxing in Bazel. This is a problem. For example, a rome.json configuration file in a parent directory would create false positives.

Are you open to having the feature **opt-out? I made the feature like this because it's used by the LSP too, and this is a feature requested by **many** users. And, as far as I understood, the prettier/eslint VSCode extensions do the same.

What do you think?

**Comment by realtimetodie** [comment]
Yes of course, opt-out is fine. In Bazel, the path to the rome.json configuration path is set explicitly using the new `--config-path` option.

**Comment by realtimetodie** [comment]
Maybe auto-traversal could be disabled, when the `--option-path` option is used (`ConfigurationBasePath::FromUser(PathBuf)`). Then we would not need an opt-out option?

**Comment by ematipico** [comment]
> Maybe auto-traversal could be disabled, when the `--config-path` option is used (`ConfigurationBasePath::FromUser(PathBuf)`). Then we would not need a separate opt-out option?

That's a good suggestion! I like it!

**Review by realtimetodie — CHANGES_REQUESTED** [review]

File: crates/rome_service/src/configuration/mod.rs [hunk 3]
```diff
@@ -129,53 +133,88 @@ pub enum BasePath {
...
     let config_name = file_system.config_name();
-    let configuration_path = match base_path {
-        BasePath::Lsp(ref path) | BasePath::FromUser(ref path) => path.join(config_name),
+    let mut configuration_path = match base_path {
+        ConfigurationBasePath::Lsp(ref path) | ConfigurationBasePath::FromUser(ref path) => {
+            path.join(config_name)
+        }
         _ => PathBuf::from(config_name),
     };
+    let should_error = base_path.is_from_user();
```
[the remark this answers is not in the document]
Comment (ematipico) [reply crates/rome_service/src/configuration/mod.rs:162]: The function is a `const`, the compiler will be able to optimize it ahead

**Pull request merged**